src/process.cr

```diff
@@ -491,6 +491,116 @@ class Process
     io.close if io
   end
 
+  # Returns the **real** user id of the current process.
+  def self.uid : LibC::UidT
```

Comment:
LibC types shouldn't be exposed in public Crystal APIs. We should either always convert stuff to Int64, Int32 or whatever is appropriate.

In fact, in Go this is exposed as a String. Maybe we should even return a struct representing this where the internal representation is hidden from the user (then we can use LibC::Uidt internally), because you won't be performing math or anything else with a uid.

Reply:
Sometimes users id's need to be displayed, saved, restored, copied or compared.    For example applications that provide persistent data owned by a particular user_id not stored in a filesystem.  Or sending user_id credentials via a socket where you can't send a name.  You can't reify it to a name on the sender or receiver as the process sending or receiving may be chroot'd and not have access to the names.

There is simple math on uid's via ==.  If euid == 0 or uid == 0 privileges are elevated.  Also comparisons are sometimes done against gid == egid to detect setgid programs. 

Anyone attempting to create a user add function in pure crystal needs math operations to find the highest uid that's not nobody.

Before you criticize the user creation example which I admit I have no use for, the socket example is on my list of things to do, checking setuid/gid is needed for the app I'm porting, sockets and uid maintained data are also part of this app.  I haven't submitted a PR for socket as that seems quite distant in getting approved.  

My point being various applications need the actual id's, not a abstract return type with hidden values. There's a reason that [go, nim, node, python, ruby] along with various other programming languages include getuid functions by default.

Reply:
The Go API used strings.

Reply:
How do I safely convert from `"0"` to `LibC::UidT`?  Is there an `IntX.from_s`?

Reply:
I was almost finished with converting to String's with the exception of the above problem when I decided to read the documentation.

The documentation claims it's an `int`. [https://golang.org/pkg/os/#Getegid](https://golang.org/pkg/os/#Getegid)
```
func Geteuid() int
Geteuid returns the numeric effective user id of the caller.
On Windows, it returns -1. 
```

Actual code: [https://golang.org/src/os/proc.go](https://golang.org/src/os/proc.go)
```
// Geteuid returns the numeric effective user id of the caller.

// On Windows, it returns -1.

func Geteuid() int { return syscall.Geteuid() }
```

Also windows doesn't have a concept of effective vs real vs saved user or group id's.

Personally I'd prefer to keep these as Int's as that's what everyone using them expects them to be and if someone does what they've been taught in every unix programming book to check for root by using `if euid == 0` it will actually work instead of creating security holes.

Reply:
Sorry to dig this up from the grave but I think it deserves to be said.

>> Personally I'd prefer to keep these as Int's as that's what everyone using them expects them to be

I think @didactic-drunk is correct. Anyone using this is expecting integers. Using strings just adds a more trouble to its use for the end user. More trouble means more screw ups more screw-ups mean more bugs. Seeing as these would be during runtime thats even more troublesome. Ideally this is what you try to avoid troublesome runtime things with frameworks/libraries. If it isn't supported on a OS the API shouldn't be available on that OS, throwing a compile time error. Requiring the user to support OS specific features on any given OS is to be expected.

Reply:
Maybe the difference in opinions is one of trying to create portable API (@asterite) and trying to use features we know aren't portable like `setuid`/`chroot` and would like the program to break at compile time if it can't handle the operation.

When I want a Windows `SID`, I'll ask for one and not expect it to work on POSIX.  Likewise I won't bother with `getgid` on Windows. 

100% portable applications are a nice goal, but I'm not making 100% portable applications.  I'm making POSIX applications that expect to use POSIX calls for security.  Otherwise I can't use crystal.

I could use shards but...

Maintaining shards is hard for these types of calls as the definitions vary between platforms and crystal doesn't include a c header parser.  Luckily the bindings were generated under `src/lib_c`, but using them means half the calls are in base crystal and the shard monkey patches to get access.  That seems like they could get out of sync quickly and be a maintenance nightmare.

My preference is break at compile and document as "Nonportable.  Use X if you can".